File: module-app/src/main/java/maple/expectation/service/v2/cache/RedisLikeRelationBufferAdapter.java
package maple.expectation.service.v2.cache;

import io.micrometer.core.instrument.MeterRegistry;
import java.util.Set;
import maple.expectation.infrastructure.executor.LogicExecutor;
import maple.expectation.infrastructure.queue.like.RedisLikeRelationBuffer;
import org.redisson.api.RedissonClient;

/**
 * Adapter wrapper for RedisLikeRelationBuffer to implement v2 interface.
 *
 * <p>This adapter bridges the core port interface to the deprecated v2 interface for backward
 * compatibility.
 *
 * <p><b>Phase 3 Technical Debt:</b> This adapter can be removed once all services are migrated to
 * use the core port interface directly.
 */
public class RedisLikeRelationBufferAdapter implements LikeRelationBufferStrategy {

  private final RedisLikeRelationBuffer delegate;

  public RedisLikeRelationBufferAdapter(
      RedissonClient redissonClient, LogicExecutor executor, MeterRegistry meterRegistry) {
    this.delegate = new RedisLikeRelationBuffer(redissonClient, executor, meterRegistry);
  }

  @Override
  public StrategyType getType() {
    return delegate.getType();
  }

  @Override
  public Boolean addRelation(String accountId, String targetOcid) {
    return delegate.addRelation(accountId, targetOcid);
  }

  @Override
  public Boolean exists(String accountId, String targetOcid) {
    return delegate.exists(accountId, targetOcid);
  }

  @Override
  public String buildRelationKey(String accountId, String targetOcid) {
    return delegate.buildRelationKey(accountId, targetOcid);
  }

  @Override
  public String[] parseRelationKey(String relationKey) {
    return delegate.parseRelationKey(relationKey);
  }

  @Override
  public Boolean removeRelation(String accountId, String targetOcid) {
    return delegate.removeRelation(accountId, targetOcid);
  }

  @Override
  public Set<String> fetchAndRemovePending(int limit) {
    return delegate.fetchAndRemovePending(limit);
  }

  @Override
  public int getRelationsSize() {
    return delegate.getRelationsSize();
  }

  @Override
  public int getPendingSize() {
    return delegate.getPendingSize();
  }

  @Override
  public Boolean existsInUnliked(String accountId, String targetOcid) {
    return delegate.existsInUnliked(accountId, targetOcid);
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/queue/like/RedisLikeRelationBufferAdapter.kt
package maple.expectation.infrastructure.queue.like

import io.micrometer.core.instrument.MeterRegistry
import maple.expectation.core.port.out.LikeRelationBufferStrategy
import maple.expectation.infrastructure.executor.LogicExecutor
import maple.expectation.infrastructure.executor.TaskContext
import org.redisson.api.RedissonClient

/**
 * Redis 기반 Like Relation Buffer 어댑터
 */
class RedisLikeRelationBufferAdapter(
    private val redissonClient: RedissonClient,
    private val executor: LogicExecutor,
    private val meterRegistry: MeterRegistry
) : LikeRelationBufferStrategy {

    companion object {
        private const val RELATIONS_KEY = "like:relations"
        private const val PENDING_KEY = "like:relations:pending"
        private const val UNLIKED_KEY = "like:relations:unliked"
    }

    override fun getType(): LikeRelationBufferStrategy.StrategyType =
        LikeRelationBufferStrategy.StrategyType.REDIS

    override fun addRelation(accountId: String, targetOcid: String): Boolean? {
        return executor.executeOrDefault(
            {
                val key = buildRelationKey(accountId, targetOcid)
                val relations = redissonClient.getSet<String>(RELATIONS_KEY)
                relations.add(key)
            },
            null,
            TaskContext.of("RedisLikeRelationBuffer", "AddRelation", accountId)
        )
    }

    override fun exists(accountId: String, targetOcid: String): Boolean? {
        return executor.executeOrDefault(
            {
                val key = buildRelationKey(accountId, targetOcid)
                val relations = redissonClient.getSet<String>(RELATIONS_KEY)
                relations.contains(key)
            },
            null,
            TaskContext.of("RedisLikeRelationBuffer", "Exists", accountId)
        )
    }

    override fun removeRelation(accountId: String, targetOcid: String): Boolean? {
        return executor.executeOrDefault(
            {
                val key = buildRelationKey(accountId, targetOcid)
                val relations = redissonClient.getSet<String>(RELATIONS_KEY)
                relations.remove(key)
            },
            null,
            TaskContext.of("RedisLikeRelationBuffer", "RemoveRelation", accountId)
        )
    }

    override fun fetchAndRemovePending(limit: Int): Set<String> {
        return executor.executeOrDefault(
            {
                val pending = redissonClient.getSet<String>(PENDING_KEY)
                val resultSet = mutableSetOf<String>()
                var count = 0
                val iterator = pending.iterator()
                while (iterator.hasNext() && count < limit) {
                    resultSet.add(iterator.next())
                    iterator.remove()
                    count++
                }
                resultSet.toSet()
            },
            emptySet(),
            TaskContext.of("RedisLikeRelationBuffer", "FetchAndRemovePending", limit.toString())
        ) ?: emptySet()
    }

    override fun buildRelationKey(accountId: String, targetOcid: String): String =
        "$accountId:$targetOcid"

    override fun parseRelationKey(relationKey: String): Array<String> =
        relationKey.split(":").toTypedArray()

    override fun existsInUnliked(accountId: String, targetOcid: String): Boolean? {
        return executor.executeOrDefault(
            {
                val key = buildRelationKey(accountId, targetOcid)
                val unliked = redissonClient.getSet<String>(UNLIKED_KEY)
                unliked.contains(key)
            },
            null,
            TaskContext.of("RedisLikeRelationBuffer", "ExistsInUnliked", accountId)
        )
    }

    override fun getRelationsSize(): Int {
        return executor.executeOrDefault(
            { redissonClient.getSet<String>(RELATIONS_KEY).size },
            0,
            TaskContext.of("RedisLikeRelationBuffer", "GetRelationsSize", "")
        )
    }

    override fun getPendingSize(): Int {
        return executor.executeOrDefault(
            { redissonClient.getSet<String>(PENDING_KEY).size },
            0,
            TaskContext.of("RedisLikeRelationBuffer", "GetPendingSize", "")
        )
    }
}
